src/register.cpp

```diff
@@ -46,6 +46,41 @@ using namespace duckdb;
 	}
 }
 
+
+
+unique_ptr<TableRef> duckdb::EnvironmentScanReplacement(duckdb::ClientContext &context, const std::string &table_name, duckdb::ReplacementScanData *data_p) {
+	auto &data = (ReplacementDataDBWrapper &)*data_p;
+	auto db_wrapper = data.wrapper;
+
+	auto table_name_symbol = Rf_install(table_name.c_str());
+	SEXP df;
+	SEXP rho = R_GetCurrentEnv();
+	while(rho != R_EmptyEnv) {
+		df = Rf_findVarInFrame3(rho, table_name_symbol, TRUE);
+		if (df != R_UnboundValue) {
+			break;
+		}
+		rho = ENCLOS(rho);
+	}
+	if (!df) {
+		return nullptr;
+	}
+	if (TYPEOF(df) == PROMSXP) {
+		df = Rf_eval(df, rho);
```

Comment:
cpp11 has a global protection mechanism, but how to control the lifetime?

Reply:
indeed lifetime is a problematic issue. For example, once someone creates a view of this and then deletes the df and gc runs we might be in trouble.

Reply:
IIUC, support for lifetime control has landed via duckdb/duckdb#11761, which is why we now have conflicts here?

Reply:
We already had support for lifetime control, but it was broken in a couple ways

We use the `ExternalDependency` class for lifetime control, but this wasn't being kept alive long enough, which this PR aimed to address.

We changed the prototype of the replacement scan, we changed the `const string &table_name` into `ReplacementScanInput`

I have a follow up PR scheduled that removes the newly added `table_ref &TableRef` from this struct though
Instead we will collect replacement scans inside QueryRelation::Bind (virtual overload) and push these replacement scans as CTEs onto the parsed QueryNode stored inside the QueryRelation.

Reply:
Thanks, Thijs. May I get back to you for help with lifetime control after that follow-up PR has made it into the R package? I don't find that PR in https://github.com/duckdb/duckdb/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+author%3ATishj.

Reply:
@Tishj: how would I go about controlling the lifetime of the bound object in this PR?